### DISCUSSION TITLE
How to apply displacement to specific nodes of a TetrahedralMesh?

### DISCUSSION THREAD

**Discussion by AlanConnorChen** [sofa-framework/sofa#5179] [Write / Run a simulation] [answered]
Hello,

I want to apply displacements (which can also be seen as external forces) to some specific 3D points(with their indices provided) of a Tetrahedral Mesh and model the entire Tetrahedral Mesh as a HyperElastic FEM Model, such as the ones listed here: [tetrahedronhyperelasticityfemforcefield](https://sofa-framework.github.io/doc/components/solidmechanics/fem/hyperelastic/tetrahedronhyperelasticityfemforcefield/)

The goal is to set the target positions for the deformation of these 3D points and obtain the deformed model. The displacement vectors are shown in the figure below.
![Displacement_Vector](https://github.com/user-attachments/assets/5f032653-6ed6-4e73-bb7d-f58697ab1ece)

Here are my questions:

1. To achieve the above goal, what resources should I refer to? I noticed the documentation here: [pointprojectiveconstraint](https://sofa-framework.github.io/doc/components/constraint/projective/pointprojectiveconstraint/) Can this help me achieve the desired result?

2. Additionally, I’m facing another issue where I can’t import my own .vtk file after I drag my .scn file into the gui. The SOFA interface appears blank, but there are no error messages in the terminal. My screenshot and console logs are as follows. 
![sofa_console_log](https://github.com/user-attachments/assets/4e5493ad-bddd-4ef6-b74a-62d78ff405c3)
[console_log.txt](https://github.com/user-attachments/files/18250898/console_log.txt)

The .scn file and .vtk file I’m using are available in the [google drive.](https://drive.google.com/drive/folders/1Mqxs_THf8B8hXv06yl5SFi26PzKJ4PRK?usp=sharing) Thanks in advance.

**Comment by bakpaul** [comment, marked as answer]
Hello

This is fairly easy to do. You can use `BilateralLagrangianConstraint` to impose the position of a subset of points of the model. See here an example : https://github.com/sofa-framework/sofa/blob/master/examples/Component/Constraint/Lagrangian/BilateralLagrangianConstraint_NNCG.scn

In this example all of the objects are simulated, but one of the two inputed object can be only a node containing a `MechanicalState`that contains the target points without any solvers of mechanics attached. Make sure to put this node in the rootNode, not as a child of any of the node of your liver.  

Tell me how it goes !

**Reply by AlanConnorChen** [reply]
Hello

Thank you for your reply, sorry for the late response. I have already opened the example file and observed its simulation effect.
![Screenshot from 2025-01-23 15-23-26](https://github.com/user-attachments/assets/c49c076a-70c1-44ef-98f5-c71245e0f573)

As for the `BilateralLagrangianConstraint` you mentioned, I am still thinking about how to process and import my own VTK file, and I am currently still in the exploration process.

Thank you for your reply!
